### PR TITLE
Fix #713 - node sourcemap troubleshooting

### LINKED ISSUE #713: Add info on Node debug troubleshooting
Look at common node-specific issues and add a few notes about them. Especially sourcemap issues and breakpoints not binding. There is very little official info out there, beyond github issues.

Example from chrome-debug: https://github.com/Microsoft/vscode-chrome-debug#troubleshooting

### PR DESCRIPTION
**Pull request by roblourens**
Please temporarily forget everything you know about sourcemaps, read this, and tell me if it could be helpful.

It might be too much information. On one hand, detailed info on how this process works may help some users help themselves. On the other hand, I don't want to scare users away by making this seem more complicated than it is. Most of the time, debugging with source maps just works. The rest of the time, it's usually one of the first 3 bullet points. 

The important ideas in the paragraphs before the bulleted list are
* We have to get the .js first before we can map .js <-> .ts
* The paths in the source map file have to be correct

If this should be a detailed explanation of everything that happens, it could be expanded. If it should be more of a basic troubleshooting list, it could be cut.

I'm not mentioning the `trace` option since it's not quite official and we should decide how to standardize between node and node2 before documenting it.

### CODE REVIEW

File: docs/editor/debugging.md [hunk 1]
```diff
@@ -319,6 +319,27 @@ Source maps can be generated with two kinds of inlining:
...
+
+When you build your `app.ts` file in TypeScript with source maps enabled, it either produces an `app.js.map` file, or a source map inlined as a base64-encoded string in a comment at the bottom of the `app.js` file. To find the `.ts` files associated with this map, the debug adapter looks at two properties in the source map, `sources`, and `sourceRoot`. `sourceRoot` is optional - if present, it is prepended to each path in `sources`, which is an array of paths. The result is an array of absolute or relative paths to `.ts` files. Relative paths are resolved relative to the source map.
+
+Finally, the debug adapter searches for the full path of `app.ts` in this resulting list of `.ts` files. If there's a match, it has found the source map file to use when mapping `app.ts` to `app.js`. If there is no match, then it can't bind the breakpoint, and it will turn gray.
+
+Here are some things to try when your breakpoints turn gray.
+* First and most important: do you have `"sourceMaps": true` in your launch config?
```
Comment (isidorn) [inline docs/editor/debugging.md]: No need to put first and most important here. I mean it is already first in the list :)

File: docs/editor/debugging.md [hunk 2]
```diff
@@ -319,6 +319,27 @@ Source maps can be generated with two kinds of inlining:
 
 VS Code supports both the **inlined source maps** and the **inlined source**.
```
Comment (isidorn) [inline docs/editor/debugging.md]: I think using the term `launch.json` is clearer than launch config. Also I feel that launch.json is what we usually use in other places.

File: docs/editor/debugging.md [hunk 3]
```diff
@@ -319,6 +319,27 @@ Source maps can be generated with two kinds of inlining:
...
+* Try the new experimental `node2` debug adapter. It can handle some more of the more complex source map cases.
+* Are the `sourceRoot` and `sources` properties in your source map correct? Can they be combined to get the correct path to the `.ts` file?
+* Are you using Webpack? By default, it outputs paths with a `webpack:///` prefix, which the debug adapter can't resolve. You can change this in your Webpack config with the `devtoolModuleFilenameTemplate` option, or try using `node2`, which provides some extra options for resolving these paths.
+* Have you opened the folder in VS Code with the incorrect case? It's possible to open folder `foo/` from the command line like `code FOO` in which case source maps may not be resolved correctly.
+* Try adding a `debugger` statement. If it breaks into the `.ts` file there, but breakpoints at that spot don't bind, that is useful information to include with a Github issue. If it breaks into a `.ts` file that says "read-only inline content from source map" next to it in Open Editors, then that's a sure sign that the source is inlined in the source map, and the path in the source map is incorrect.
+
```
Comment (isidorn) [inline docs/editor/debugging.md]: All of this bullet points should be copied into a fake stackOverflow question and answer. Since a lot of users will first seach there.
Example me asking and ben answering is the first result :D https://www.google.rs/?gws_rd=cr&ei=nQtRWJyWEoXPa9LGirgB#q=how%20to%20disable%20tabs%20vscode

File: docs/editor/debugging.md [hunk 4]
```diff
@@ -319,6 +319,27 @@ Source maps can be generated with two kinds of inlining:
...
+Here are some things to try when your breakpoints turn gray.
+* First and most important: do you have `"sourceMaps": true` in your launch config?
+* Did you build with source maps enabled? Are there `.js.map` files, or inlined source maps in your `.js` files?
+* Did you set the `outFiles` property in your launch config? It should be a glob pattern for an absolute path that matches your `.js` files.
+* Try the new experimental `node2` debug adapter. It can handle some more of the more complex source map cases.
+* Are the `sourceRoot` and `sources` properties in your source map correct? Can they be combined to get the correct path to the `.ts` file?
+* Are you using Webpack? By default, it outputs paths with a `webpack:///` prefix, which the debug adapter can't resolve. You can change this in your Webpack config with the `devtoolModuleFilenameTemplate` option, or try using `node2`, which provides some extra options for resolving these paths.
```
Comment (isidorn) [inline docs/editor/debugging.md]: It is not 100% clear how to use the new `node2` adapter. Maybe mention to simple put it as a type in your launch.json config. Or add a link to the section where node2 is explained.

File: docs/editor/debugging.md [hunk 5]
```diff
@@ -319,6 +319,27 @@ Source maps can be generated with two kinds of inlining:
...
+* Have you opened the folder in VS Code with the incorrect case? It's possible to open folder `foo/` from the command line like `code FOO` in which case source maps may not be resolved correctly.
+* Try adding a `debugger` statement. If it breaks into the `.ts` file there, but breakpoints at that spot don't bind, that is useful information to include with a Github issue. If it breaks into a `.ts` file that says "read-only inline content from source map" next to it in Open Editors, then that's a sure sign that the source is inlined in the source map, and the path in the source map is incorrect.
```
Comment (isidorn) [inline docs/editor/debugging.md]: The last point feels like too much detail that only a person who is filling a bug should know. I would remove it - unless we get a special page only for source maps.
Or I would rephrase it to the following: "Still not working, file an issue lalala and then mention this useful detail"

File: docs/editor/debugging.md [hunk 6]
```diff
@@ -319,6 +319,27 @@ Source maps can be generated with two kinds of inlining:
 
 VS Code supports both the **inlined source maps** and the **inlined source**.
 
+#### JavaScript Source Map Tips
+
```
Comment (isidorn) [inline docs/editor/debugging.md]: I would like to see an example of how source maps need to be setup.
Maybe create a public git hub repositore as a sample and then just link it from here.

Comment (roblourens) [inline docs/editor/debugging.md]: We do have the sample projects, I can link them

File: docs/editor/debugging.md (hunk 5 above)
Comment (roblourens) [inline docs/editor/debugging.md]: It does feel weird, it's also the only actually concrete piece of info. I'll make that change tho

File: docs/editor/debugging.md (hunk 3 above)
Comment (roblourens) [inline docs/editor/debugging.md]: There are already a bunch of very not helpful SO posts. Not sure if I can cut through the noise :)

File: docs/editor/debugging.md (hunk 6 above)
Comment (roblourens) [inline docs/editor/debugging.md]: Do you know what these sample projects are actually used for? I don't see where they are linked. And they might need a little work - https://github.com/Microsoft/vscode-samples/tree/master/node-express-typescript

Comment (isidorn) [inline docs/editor/debugging.md]: No clue. @chrisdias might know

Comment (chrisdias) [inline docs/editor/debugging.md]: these were very early samples of basic express apps that were fully configured to get the best out of VS Code. I don't think we reference them anywhere.

Comment (roblourens) [inline docs/editor/debugging.md]: I think that a really simple example of sourcemap debugging isn't that helpful because it's usually easy to set up. Problems come in with more complicated setups, and those are all different.

File: docs/editor/debugging.md (hunk 4 above)
Comment (roblourens) [inline docs/editor/debugging.md]: Oh yeah I was going to do that but that section is only in vnext. I filed #753 to remind myself.

**Comment by gregvanl** [comment]
Oops, I got merge-happy and didn't realize this was assigned to Isi and Andre.
